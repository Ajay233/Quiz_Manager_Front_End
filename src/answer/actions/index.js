import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
