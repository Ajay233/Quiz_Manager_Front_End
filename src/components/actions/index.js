export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
