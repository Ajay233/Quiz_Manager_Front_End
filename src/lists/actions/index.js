import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
