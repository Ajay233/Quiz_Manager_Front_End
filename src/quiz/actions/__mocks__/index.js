export const setQuiz = jest.fn()
export const createQuiz = jest.fn( () => { return {type: "SET_QUIZ", payload:""} })
export const updateQuiz = jest.fn( () => { return {type: "SET_QUIZ", payload:""} })
export const updateQuizStatus = jest.fn()
