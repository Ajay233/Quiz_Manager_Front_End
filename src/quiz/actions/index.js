import { get, getUsingParams } from '../../axiosRequests/requests';

export const setQuizes = (quizes) => {
  return {
    type: "SET_QUIZES",
    payload: quizes
  }
}

export const getAllQuizes = (endpoint, data, token) => {
  return (dispatch) => {
    get(endpoint, data, token).then((response) => {
      console.log(response)
      dispatch({
        type: "GET_ALL_QUIZES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
    });
  }
}

export const getQuizSearchResults = (endpoint, data, token, setNotification, errorMsg) => {
  return (dispatch) => {
    getUsingParams(endpoint, data, token).then((response) => {
      dispatch({
        type: "SET_QUIZ_SEARCH_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response);
      setNotification(errorMsg, "error", true);
    })
  }
}
