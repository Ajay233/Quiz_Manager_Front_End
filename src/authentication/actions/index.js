import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
