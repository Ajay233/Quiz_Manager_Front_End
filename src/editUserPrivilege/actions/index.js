import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
