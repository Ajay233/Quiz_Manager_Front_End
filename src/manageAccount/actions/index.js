import { put } from '../../axiosRequests/requests'
import { setUserData } from '../../authentication/actions'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const updateUser = (data, jwt) => {
  const emailChangeMsg = "Profile information updated - Your email was changed so your access has been restriced until your email is verified"
  const errorMsg = "Error updating your profile data, please check the details you provided and try again";
  const emailTaken = "An account with that email already exists"
  return (dispatch) => {
    return put('users/update', data, jwt).then((response) => {
      if(response.data.jwt === undefined){
        dispatch(setUserData(response.data, jwt))
        dispatch(setNotification("Profile information updated", "success", true))
      } else {
        dispatch(setUserData(response.data.user, response.data.jwt))
        dispatch(setNotification(emailChangeMsg, "success", true))
      }
    }).catch((error) => {
      console.log(error)
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else if(error.response.data === emailTaken){
        dispatch(setNotification(emailTaken, "error", true))
      } else {
        dispatch(setNotification(errorMsg, "error", true))
      }
    });
  }
}
