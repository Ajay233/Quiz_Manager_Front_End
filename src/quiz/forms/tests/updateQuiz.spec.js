import React from 'react'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import UpdateQuizForm from '../updateQuiz'
import { mount } from 'enzyme'
import mockAxios from 'jest-mock-axios'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { render, fireEvent, cleanup } from '@testing-library/react'
import history from '../../../history'
import { setQuiz } from '../../actions'
import { setNotification } from '../../../notifications/actions'
import { sessionExpired } from '../../../utils/session'

jest.mock('../../actions')
jest.mock('../../../notifications/actions')
jest.mock('../../../utils/session')

const mockStore = configureStore({})

describe("UpdateQuiz Form", () => {

  let store;

  beforeEach(() => {
    store = mockStore({
      quiz: {
        id: 1,
        name: "testName",
        description: "testDescription",
        category: "testCategory",
        status: "DRAFT"
      },
      lists: {
        categories: ["testCategory", "testCategor2", "testCategor3"]
      },
      userData: {
        id: 1,
        forename: "testForename",
        surname: "testSurname",
        email: "testEmail",
        permission: "USER",
        verified: "true",
        jwt: "testJwt",
        loggedIn: true
      }
    })
  })

  afterEach(() => {
    mockAxios.reset()
  })

  it("should call action creators on submit success", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuizForm />
        </Router>
      </Provider>
    )

    const requestResponse = {
      data: {
        id: 1,
        name: "updatedName",
        description: "updatedDescription",
        category: "updatedCategory",
        status: "READY"
      }
    }

    wrapper.find('form').simulate('submit')
    mockAxios.mockResponse(requestResponse)
    expect(mockAxios.put).toHaveBeenCalledTimes(1)
    expect(setQuiz).toHaveBeenCalledTimes(1)
    expect(setQuiz).toHaveBeenCalledWith(requestResponse.data)
  })

  it("should call sessionExpired on submit error status 403", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuizForm />
        </Router>
      </Provider>
    )

    const errorResponse = {
      response: {
        status: 403
      }
    }

    wrapper.find('form').simulate('submit')
    mockAxios.mockError(errorResponse)
    expect(mockAxios.put).toHaveBeenCalledTimes(1)
    expect(sessionExpired).toHaveBeenCalledTimes(1)
  })

  it("should call setNotification on any other submit error", () => {
    const wrapper = mount(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuizForm />
        </Router>
      </Provider>
    )

    const errorResponse = {
      response: {
        data: "NOT FOUND",
        status: 404
      }
    }

    wrapper.find('form').simulate('submit')
    mockAxios.mockError(errorResponse)
    expect(mockAxios.put).toHaveBeenCalledTimes(1)
    expect(setNotification).toHaveBeenCalledTimes(1)
  })

  it("should call updateStatus when the update status button is clicked", () => {

    const updateStatus = jest.fn()

    const wrapper = render(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuizForm
            updateStatus={updateStatus}
          />
        </Router>
      </Provider>
    )

    fireEvent.click(wrapper.getByTestId("updateStatus-button"))
    expect(updateStatus).toHaveBeenCalledTimes(1)
  })

  it("should match the snapshot", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <UpdateQuizForm />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

})
