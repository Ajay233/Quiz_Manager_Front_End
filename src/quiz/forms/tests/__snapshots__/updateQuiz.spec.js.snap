// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateQuiz Form should match the snapshot 1`] = `
<div
  className=""
>
  <div
    className=""
  >
    <form
      className="form"
      onSubmit={[Function]}
    >
      <div>
        <label>
          Quiz name:
        </label>
        <input
          className="inputBox"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="testName"
          value=""
        />
      </div>
      <div>
        <label>
          Quiz description:
        </label>
        <textarea
          className="inputBox"
          name="description"
          onBlur={[Function]}
          onChange={[Function]}
          onDragStart={[Function]}
          onDrop={[Function]}
          onFocus={[Function]}
          placeholder="testDescription"
          value=""
        />
      </div>
      <div>
        <label>
          Quiz category:
        </label>
        <div>
          <select
            className="inputBox select-medium"
            name="category"
            onBlur={[Function]}
            onChange={[Function]}
            onDragStart={[Function]}
            onDrop={[Function]}
            onFocus={[Function]}
            placeholder="testCategory"
            value=""
          >
            <option
              value="testCategory"
            >
              testCategory
            </option>
            <option
              value="testCategor2"
            >
              testCategor2
            </option>
            <option
              value="testCategor3"
            >
              testCategor3
            </option>
          </select>
        </div>
      </div>
      <div
        className="buttonGroup"
      >
        <button
          className="submit"
        >
          <i
            className="far fa-save"
          />
           Save Changes
        </button>
        <button
          className="delete"
          data-testid="delete-quiz-button"
        >
          <i
            className="fas fa-trash-alt"
          />
           Delete Quiz
        </button>
        <button
          className="save"
          data-testid="updateStatus-button"
        >
          <i
            className="far fa-check-circle"
          />
           Mark as Ready
        </button>
      </div>
    </form>
  </div>
  <div
    className=""
  />
</div>
`;
